Sources/SKSupport/LineTable.swift

```diff
@@ -172,29 +111,81 @@ extension LineTable {
   /// - parameter fromOff: Starting UTF-16 column offset (zero-based).
   /// - parameter utf16Length: The number of UTF-16 code units to replace.
   /// - parameter replacement: The new text for the given range.
-  @inlinable
   mutating public func replace(
     fromLine: Int,
     utf16Offset fromOff: Int,
     utf16Length: Int,
     with replacement: String)
   {
-    let endOff = self[fromLine].utf16Offset + fromOff + utf16Length
-    let endLine = self[utf16Offset: endOff]
-
-    self.replace(fromLine: fromLine, utf16Offset: fromOff, toLine: endLine.index, utf16Offset: endOff - endLine.utf16Offset, with: replacement)
+    let start = content.utf16.index(impl[fromLine], offsetBy: fromOff)
+    let end = content.utf16.index(start, offsetBy: utf16Length)
+    let (toLine, toOff) = lineAndUTF16ColumnOf(end, fromLine: fromLine)
+    self.replace(fromLine: fromLine, utf16Offset: fromOff, toLine: toLine, utf16Offset: toOff, with: replacement)
   }
 }
 
-// Note: This is copied from the stdlib.
-// Used to calculate a running count. For non-BMP scalars, it's important if the
-// prior code unit was a leading surrogate (validity).
-private func _utf8Count(_ utf16CU: UInt16, prev: UInt16) -> Int {
-  switch utf16CU {
-  case 0..<0x80: return 1
-  case 0x80..<0x800: return 2
-  case 0x800..<0xDC00: return 3
-  case 0xDC00..<0xE000: return UTF16.isLeadSurrogate(prev) ? 1 : 3
-  default: return 3
+extension LineTable {
+
+  // MARK: - Position translation
+
+  /// Returns `String.Index` of given logical position.
+  ///
+  /// - parameter line: Line number (zero-based).
+  /// - parameter utf16Column: UTF-16 column offset (zero-based).
+  @inlinable
+  public func stringIndexOf(line: Int, utf16Column: Int) -> String.Index? {
+    guard line < count else {
+      // Line out of range.
+      return nil
+    }
+    let lineSlice = self[line]
+    guard utf16Column <= content.utf16.distance(from: lineSlice.startIndex, to: lineSlice.endIndex) else {
+      // Column out of range.
+      return nil
+    }
+    return content.utf16.index(lineSlice.startIndex, offsetBy: utf16Column)
```

Comment:
Hm, I've totally forgotten `index(_:offsetBy:limitedBy:)`. line 142-146 can be just:
```swift
  return content.utf16.index(lineSlice.startIndex, offsetBy: utf16Column, limitedBy: lineSlice.endIndex)
```